Code/src/main/java/nl/utwente/group10/ui/components/lines/Connection.java

```diff
@@ -138,162 +162,138 @@ public boolean tryAddAnchor(ConnectionAnchor anchor, boolean overrideExisting, b
         if (newAnchor instanceof OutputAnchor) {
-            startAnchor.ifPresent(a -> removeListeners(a));
+            startAnchor.ifPresent(a -> disconnect(a));
             startAnchor = Optional.of((OutputAnchor) newAnchor);
         } else if (newAnchor instanceof InputAnchor) {
-            endAnchor.ifPresent(a -> removeListeners(a));
+            endAnchor.ifPresent(a -> disconnect(a));
             endAnchor = Optional.of((InputAnchor) newAnchor);
+        } else {
+            throw new InvalidAnchorException();
         }
+        
+        // Add this to the anchor.
+        newAnchor.addConnection(this);
+        addListeners(newAnchor);
 
-        ConnectionCreationManager.nextConnectionState();
-        invalidateConnectionStateCascading();
+        if (isConnected()) {
+            // The ConnectionState changed.
+            ConnectionCreationManager.nextConnectionState();
+            startAnchor.get().getBlock().setConnectionState(ConnectionCreationManager.getConnectionState());
+            endAnchor.get().getBlock().setConnectionState(ConnectionCreationManager.getConnectionState());
+        }
+        invalidateAnchorPositions();
     }
 
     /**
-     * Adds the listeners this Connections needs to keep its visual
-     * representation up-to-date to the given anchor.
+     * Sets the free ends (empty anchors) to the specified position.
+     * 
+     * @param point
+     *            Coordinates local to the Line's parent.
      */
-    private void addListeners(ConnectionAnchor anchor) {
-        anchor.layoutXProperty().addListener(this);
-        anchor.layoutYProperty().addListener(this);
-        anchor.getBlock().layoutXProperty().addListener(this);
-        anchor.getBlock().layoutYProperty().addListener(this);
+    public void setFreeEnds(Point2D point) {
+        if (!startAnchor.isPresent()) {
+            setStartPositionParent(point);
+        }
+        if (!endAnchor.isPresent()) {
+            setEndPositionParent(point);
+        }
     }
-
+    
     /**
-     * Removes the listeners this Connections needed to keep its visual
-     * representation up-to-date from the given anchor.
+     * Listener method that can be attached to a BooleanProperty in order to
+     * update the error state based on that property.
      */
-    private void removeListeners(ConnectionAnchor anchor) {
-        anchor.layoutXProperty().removeListener(this);
-        anchor.layoutYProperty().removeListener(this);
-        anchor.getBlock().layoutXProperty().removeListener(this);
-        anchor.getBlock().layoutYProperty().removeListener(this);
+    private void checkErrorListener(ObservableValue<? extends Boolean> value, Boolean oldValue, Boolean newValue) {
+        ObservableList<String> styleClass = this.getStyleClass();
+        if (newValue) {
+            styleClass.removeAll("error");
+            styleClass.add("error");
+        } else {
+            styleClass.removeAll("error");
+        }  
```

Comment:
Could restructure this if to always removeAll("error") and only add it back if newValue.